### PR TITLE
2018Q4: add engineering OKRs

### PR DESCRIPTION
**Pull request by smola**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,114 @@
...
+
+- [ ] Incremental PGA updates [APPS]
+- [ ] Automate PGA generation [APPS][INFRA]
+- [ ] Release PGA v2 [APPS]
+
+## Bring 'Assisted Code Review' to developers in the open-source community
+
+### Get lookout production ready [APPS]
+
+- [ ] Extend lookout end user documentation
```
Comment (bzz) [inline README.md]: would it worth to be specific here, same as in borges below, and mention that docs should be using Gitbook? that would make sense to me

File: README.md [hunk 2]
```diff
@@ -1 +1,114 @@
 # OKRs 2018 Q4
+
+## Bring 'Code as Data' to market with enterprises
+
+### Ensure gitbase interoperability with 3rd party tools [DR]
+
+- [ ] Drivers (JS, PHP, Python, Ruby, Go, Java, .NET)
```
Comment (ajnavarro) [inline README.md]: that points are in priority order?

File: README.md [hunk 3]
```diff
@@ -1 +1,114 @@
...
+- [ ] mysqldump
+- [ ] grafana
+- [ ] BI (Tableau, PowerBI, pick one)
+
+### Get gitbase enterprise ready [DR]
+
+- [ ] Authentication and authorization abstractions with basic MySQL auth implementation
```
Comment (ajnavarro) [inline README.md]: I cannot understand what is required here. Roles? read only users? ...

File: README.md [hunk 4]
```diff
@@ -1 +1,114 @@
...
+
+### Get web interfaces enterprise ready [APPS]
+
+- [ ] gitbase-web SSO (protocol TBD)
+- [ ] bblfsh-web SSO (protocol TBD)
+
+### Gitbase research [DR]
+
+- [ ] Support updates in backing data (indexes, etc)
+- [ ] Usage of caches for UAST and others
```
Comment (ajnavarro) [inline README.md]: this is already done. We are using a global object cache, languages cache and UAST cache.

**Review by dennwc — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -1 +1,114 @@
...
+
+### Reach a stable version on borges & rovers
+
+#### Refactor to go-borges library, to reuse borges logic in other projects (i.e. Lookout) [DR]
+
+- [ ] Define a public API for that library studying actual use cases
+
+#### Improve borges documentation [DR]
+
+- [ ] Gitbook integration
```
Comment (ajnavarro) [inline README.md]: Already done

File: README.md (hunk 2 above)
Comment (smola) [reply README.md]: Not exactly. We could break them in two tiers for priorities:
- Drivers for in-house languages (Python, Go, Java).
- mysqldump
- grafana
- Tableau
- Drivers for other popular languages (JavaScript, Ruby, .NET, PHP).

File: README.md (hunk 3 above)
Comment (smola) [reply README.md]: @ajnavarro  We'll discuss during the design process, but the minimum implementation would be accepting authentication with user+pass from MySQL protocol, with a defined user+password in a file, and an authorizer that authorizes everything for any authenticated user.

However, the design should consider future integration with LDAP and Kerberos, as well as authorizing read/write operations independently.

File: README.md (hunk 4 above)
Comment (smola) [reply README.md]: Actually this is about precomputing caches and persisting them.

File: README.md [hunk 6]
```diff
@@ -1 +1,112 @@
...
+
+### Get bblfsh ready for wide language support of our use cases [LA]
+
+- Finish migration to v2 for all supported client languages (JVM, Python, Go)
+- Deprecate v1
+- Fix positions in all alpha+ drivers
+
+### Have an available demo cluster [INFRA]
+
+- [ ] Deploy a demo cluster for 'Code as Data' that we can use for demos and 3rd parties
```
Comment (rporres) [inline README.md:47]: I assume this is a separate cluster, not the pipeline cluster, isn't it? At some point @mcuadros wasn't very interested in this idea...

File: README.md [hunk 7]
```diff
@@ -1 +1,112 @@
...
+
+### Improve borges archiving performance
+
+- [ ] Improve borges archiving average throughput by 5x
+
+### Refactor to go-borges library, to reuse borges logic in other projects (i.e. Lookout) [DR]
+
+- [ ] Define a public API for that library studying actual use cases
+
+### Release the production kubernetes cluster [INFRA]
```
Comment (rporres) [inline README.md:113]: At some point we should add monitoring into it

File: README.md [hunk 8]
```diff
@@ -1 +1,112 @@
...
+
+- [ ] Release of the production cluster
+- [ ] Run rovers and borges on it
+- [ ] Deploy services with HA
```
Comment (rporres) [inline README.md]: Do you mean postgres and RabbitMQ?

Comment (smola) [reply README.md]: @rporres PostgreSQL, RabbitMQ, etcd

File: README.md (hunk 6 above)
Comment (smola) [reply README.md:47]: @rporres AFAIK this is hard requirement, but we may just use GKE for this.
cc @eiso

Comment (rporres) [reply README.md:47]: It's not that I don't like it, just remembering conversations around this in the past

Comment (campoy) [reply README.md:47]: This is something that I'd like to be able to use when demoing the engine with large customers.
Somewhere that is running some stable release of our tooling and with all the necessary set-up to make everything work as fast as possible.

File: README.md [hunk 9]
```diff
@@ -1 +1,112 @@
...
+
+### Get Gemini production ready [APPS]
+
+- [ ] Release a new version with the integration of function similarity
+
+## Improve the 'Code as Data' individual user experience
+
+- [ ] Move engine maintainership to apps [APPS]
+
+## Improve quality across all products [INFRA][DR]
```
Comment (campoy) [inline README.md:61]: Does this include performance monitoring?

Comment (smola) [reply README.md:61]: Yes, it's included in borges and gitbase regression tests. Other projects will follow.

File: README.md (hunk 7 above)
Comment (smola) [reply README.md:113]: Added an OKR about monitoring, let's discuss its scope next week.

File: README.md [hunk 10]
```diff
@@ -1 +1,115 @@
...
+
+- [ ] Drivers for in-house languages (Python, Go, Java).
+- [ ] mysqldump
+- [ ] grafana
+- [ ] Tableau
+- [ ] Drivers for other popular languages (JavaScript, Ruby, .NET, PHP).
+
+### Get gitbase enterprise ready [DR]
+
+- [ ] Authentication and authorization abstractions prepared for future integration with LDAP and Kerberos
```
Comment (ajnavarro) [inline README.md:15]: `authorization abstractions` means an authorization to specific repositories, per example?

File: README.md [hunk 11]
```diff
@@ -1 +1,115 @@
...
+- [ ] Audit logs (connections, queries)
+
+### Get web interfaces enterprise ready [APPS]
+
+- [ ] gitbase-web SSO (protocol TBD)
+- [ ] bblfsh-web SSO (protocol TBD)
+
+### Gitbase research [DR]
+
+- [ ] Support updates in backing data (indexes, etc)
```
Comment (ajnavarro) [inline README.md]: with `updates` you mean: detect that a repository changed using a 3rd party tool and invalidate any generated data by gitbase (and generate them again if necessary).

Comment (smola) [reply README.md]: That's right. No `UPDATE`/`INSERT`, but data changing in the backend.

File: README.md (hunk 10 above)
Comment (smola) [reply README.md:15]: No. Just to connect to gitbase and execute `read` (SELECT) or `write` (CREATE INDEX) queries.

File: README.md [hunk 12]
```diff
@@ -1 +1,115 @@
...
+
+- [ ] Test deployment for faux-customer [INFRA][ENG]
+
+### Get Gemini production ready [APPS]
+
+- [ ] Release a new version with the integration of function similarity
+
+## Improve the 'Code as Data' individual user experience
+
+- [ ] Move engine maintainership to apps [APPS]
```
Comment (smola) [inline README.md]: Discussing this with @carlosms, this does not make sense as an OKR. engine will be maintained by apps, but we need to know if there is an actual objective for engine development this quarter.

**Comment by vmarkovtsev** [comment]
@smola I am having troubles with finding the best place to insert the following ML OKRs:

1. Design an ML competition on PGAv2 + awesome-ml list distillation
2. Publish the blog post about MSR interviews https://github.com/src-d/backlog/issues/1263
3. Publish pending datasets https://github.com/src-d/backlog/issues/1310

**Comment by vmarkovtsev** [comment]
@eiso I changed "advanced analyzer alpha" OKR to "best practices analyzer PoC" because tbh we doubt that we will collect the structured diff dataset for it in Q4 - it is very hard for the ML team. As usual, no data => no ML.

**Comment by smola** [comment]
As discussed previously with @ajnavarro, we must rephrase the "Gitbase research" OKRs so that they are actual objectives, not specific solutions.

File: README.md (hunk 6 above)
Comment (rporres) [reply README.md:47]: We have to specify then the needs, especially if we need things to be ready very fast as things do not go very fast in the baremetal world. Another thing to consider is that the minimum renting time of servers in our baremetal provider is one month.

**Comment by vmarkovtsev** [comment]
Will we have hiring OKRs?

**Comment by smola** [comment]
@vmarkovtsev AFAIK not as part of the initial engineering OKRs (this PR), but maybe they are set from another side, I moved this discussion to Slack.

File: README.md (hunk 6 above)
Comment (smola) [reply README.md:47]: Started Slack discussion at infra channel about detailed requirements for the demo cluster.

File: README.md [hunk 13]
```diff
@@ -1 +1,118 @@
...
+- [ ] Extend lookout end user documentation in gitbook
+- [ ] Extend lookout developer documentation in gitbook
+- [ ] Stable Python SDK (training-aware? DD)
+- [ ] Stable Go SDK (training-aware? DD)
+
+### Release a production ready service
+
+#### Release style analyzers [ML]
+
+- [ ] Launch beta formatting analyzer - November 1st
```
Comment (mcuadros) [inline README.md]: Why we are having dates here?

Comment (vmarkovtsev) [reply README.md]: It is easier for us to plan the work in Q4 with specific dates. So they are self-claimed. I can remove if  smth is wrong.

Comment (smola) [reply README.md]: I think we can move these to your backlog/issues, since they will not matter in terms of OKR completion.

Comment (vmarkovtsev) [reply README.md]: NP

File: README.md [hunk 14]
```diff
@@ -1 +1,118 @@
...
+
+### Get web interfaces enterprise ready [APPS]
+
+- [ ] gitbase-web SSO (protocol TBD)
+- [ ] bblfsh-web SSO (protocol TBD)
+
+### Gitbase research [DR]
+
+- [ ] Support updates in backing data (indexes, etc)
+- [ ] Speed getting UAST, possibly with the usage of persistent caches for UAST and others
```
Comment (ajnavarro) [inline README.md]: Reword to `Improve queries that are using UAST functions`

File: README.md [hunk 15]
```diff
@@ -1 +1,118 @@
...
+
+## Release a new version of PGA [INFRA][APPS]
+
+- [ ] Incremental PGA updates [APPS]
+- [ ] Automate PGA generation [APPS][INFRA]
+- [ ] Release PGA v2 [APPS]
+
+## Bring 'Assisted Code Review' to developers in the open-source community
+
+### Get lookout production ready [APPS]
```
Comment (rporres) [inline README.md:76]: this also will need some infra work to finish the deployment into staging/production and make things whatever we call "production ready". I'm again thinking in monitoring the application and analyzers

File: README.md (hunk 13 above)
Comment (smola) [reply README.md]: Removed.

File: README.md (hunk 12 above)
Comment (smola) [reply README.md]: Removed. Will bring back the meat with product OKRs.

File: README.md (hunk 15 above)
Comment (smola) [reply README.md:76]: Added KR there.

File: README.md (hunk 14 above)
Comment (smola) [reply README.md]: Rephrased to `Speed up retrieving UASTs`

**Review by ajnavarro — APPROVED** [review]

**Review by carlosms — APPROVED** [review]

**Review by smacker — APPROVED** [review]

**Comment by smola** [comment]
@mcuadros I think it's now 4 commits since your last review. I think this version might now be good to merge.

File: README.md [hunk 16]
```diff
@@ -1 +1,115 @@
...
+
+- [ ] gitbase-web SSO (protocol TBD)
+- [ ] bblfsh-web SSO (protocol TBD)
+
+### Gitbase research [DR]
+
+- [ ] Initial research on supporting updates in backing data (indexes, etc)
+- [ ] Speed up retrieving UASTs
+
+### Launch gitbase-spark-connector [DR]
```
Comment (campoy) [inline README.md:29]: Is there a design document about these connectors?
Do we have an alpha version available somewhere?

Comment (smola) [reply README.md:29]: Sent links through Slack. We are working on the first version (private), which will be splitted into community/enterprise versions this quarter.

**Comment by smola** [comment]
@vmarkovtsev @mcuadros Added the local-gpu OKR from Q3, but with better explanation of purpose. There is already a design doc explaining the details about why and how.
This affects infra @rporres

**Comment by smola** [comment]
As discussed with @mcuadros:
- Brought back original gitbase research OKR for precalculated UASTs cache, since we want to try this specifically (cc @ajnavarro)
- Bring back engine maintainership OKR for apps, since we want explicit review of code, project conventions, etc as part of the maintainership transfer (cc @src-d/applications)
And merging.

**Pull request merged**